### PR TITLE
Reporting Corfu metrics to MBeans

### PR DESCRIPTION
**Pull request by StrikingThirteen**
## Overview
Description:
This PR adds the ability of reporting the currently collected metrics to MBeans. In addition some refactoring is done to improve the clarity of metrics utility.

Why should this be merged: 
- Reported metrics to MBeans can be conveniently examined and monitored live using the tools conforming to JMX api such as jConsole and jVisualVM which are already bundled with standard JDK. This enables both users and developers of Corfu to easily access Corfu metrics.
- In regards to refactoring, the changes are steps toward improving metrics infrastructure to better address the metrics collection and reporting requirement.  

Related issue(s) (if applicable): #1243 

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **9a1bf93**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1245 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/mZ278UgPDyTijkYUP0o9vRINVq88xZ9z)

File: runtime/src/main/java/org/corfudb/util/MetricsUtils.java [hunk 1]
```diff
@@ -25,46 +27,54 @@
...
+            metricsJmxReportingEnabled = Boolean.valueOf(System.getProperty(PROPERTY_JMX_REPORTING));
+            metricsLogInterval = Integer.valueOf(System.getProperty(PROPERTY_LOG_INTERVAL));
+            metricsSlf4jReportingEnabled = metricsLogInterval > 0 ? true : false;
         } catch (NumberFormatException e) {
-            log.error("Extracting metricsInterval VM property failed. " +
-                    "Metrics collection and reporting will be disabled");
-            interval = -1;
+            log.warn("Extracting metricsInterval VM property failed. " +
+                    "Reporting to corfu metrics log is disabled");
+            metricsSlf4jReportingEnabled = false;
```
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/util/MetricsUtils.java]: `metricsSlf4jReportingEnabled` is already defaulted to false. Do we need this?

File: runtime/src/main/java/org/corfudb/util/MetricsUtils.java [hunk 2]
```diff
@@ -75,42 +85,63 @@ private static void loadVmProperties() {
...
     }
 
-    public static void addCacheGauges(MetricRegistry metrics, String name, Cache cache) {
+    // If enabled, setup jmx reporting
+    private static void setupJmxReporting(MetricRegistry metrics) {
+        if (!metricsJmxReportingEnabled) return;
+
+        // This filters noisy addressSpace metrics to have a clean JMX reporting
+        MetricFilter addressSpaceFilter = (name, metric) ->
+                !name.contains(ADDRESS_SPACE_METRIC_PREFIX);
```
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/util/MetricsUtils.java]: Is it just the Address Space Metrics which needs to be filtered? Should this be configurable like the slf4j configurable filter?

File: runtime/src/main/java/org/corfudb/util/MetricsUtils.java (hunk 1 above)
Comment (StrikingThirteen) [reply runtime/src/main/java/org/corfudb/util/MetricsUtils.java]: This is for the case where loadVmProperties is called to detect the changes on System properties. Although currently not being used for re-setting the reporting properties, loadVmProperties must be able to extract properties and correctly update it in such cases.

File: runtime/src/main/java/org/corfudb/util/MetricsUtils.java (hunk 2 above)
Comment (StrikingThirteen) [reply runtime/src/main/java/org/corfudb/util/MetricsUtils.java]: That is a valid point. I decided against parametrizing it for two reasons. First: currently, the mentioned filter, filters the noisiest metrics that was making the visual representations less useful. For now it suffices. Second: this patch has the scope of enabling the reporting of current metrics to MBeans for the urgent need being able to examine them live. There is an open requirement to weed out and restructure the entire collected metrics which will affect the mechanisms for filtering but stays outside the scope of this commit.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **c22c224**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1245 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/A3lt653X8GZt3V6SX9zmKVLUKe39UqQV)

**Pull request merged**